src/main/kotlin/build/buf/intellij/module/cache/ModuleCacheV2.kt

```diff
@@ -0,0 +1,145 @@
+import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.openapi.vfs.VfsUtil
+import com.intellij.util.io.delete
+import java.io.IOException
+import java.nio.file.DirectoryNotEmptyException
+import java.nio.file.FileAlreadyExistsException
+import java.nio.file.FileSystemException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.StandardCopyOption
+import kotlin.io.path.exists
+import kotlin.io.path.isDirectory
+
+/**
+ * Module cache implementation for the v2 Buf CLI cache directory layout.
+ * This differs from v1/v3 in that the files are stored as a content-addressable storage (CAS).
+ * To work with the IDE, we need to copy the files to a file layout based on the paths within a module.
+ */
+internal class ModuleCacheV2(env: Map<String, String> = System.getenv()) : BaseModuleCache(env) {
+
+    override fun moduleDataRoot(): Path = Path.of(SYSTEM_BUF_MOD_CACHE)
+
+    override fun moduleDataPathForModuleKey(key: ModuleKey): Path {
+        // First check if we've already copied the CAS based cache to the IDE system directory
+        val fullNamePath = "${key.moduleFullName}"
+        val bufModuleCachePath = "$bufCacheDir/v2/module/$fullNamePath"
+        val modcachePath = moduleDataRoot().resolve("$fullNamePath/${key.commitID.toDashless()}")
+        try {
+            copyFromV2ModuleCacheToSystemDir(key, Path.of(bufModuleCachePath), modcachePath)
+        } catch (e: Exception) {
+            LOG.warn("failed to copy buf v2 cache module $key: $e")
+        }
+        return modcachePath
+    }
+
+    private fun copyFromV2ModuleCacheToSystemDir(
+        key: ModuleKey,
+        sourceModuleCachePath: Path,
+        modcachePath: Path,
+    ) {
+        if (modcachePath.isDirectory()) {
+            return
+        }
+        val manifest = manifestForModuleKey(key).getOrNull() ?: return
+        if (!modcachePath.parent.isDirectory()) {
+            check(modcachePath.parent.toFile().mkdirs()) { "failed to create directory: ${modcachePath.parent}" }
+        }
+        val commitTempPath = Files.createTempDirectory(modcachePath.parent, "tmp_" + key.commitID.toDashless())
+        try {
+            for (fileNode in manifest.getFileNodes()) {
+                val cacheBlobPath = sourceModuleCachePath.resolve(digestToBlobPath(fileNode.digest))
+                val modulePath = commitTempPath.resolve(fileNode.path)
+                if (!modulePath.parent.isDirectory()) {
+                    Files.createDirectories(modulePath.parent)
+                }
+                Files.copy(cacheBlobPath, modulePath)
+            }
+            try {
+                Files.move(commitTempPath, modcachePath, StandardCopyOption.ATOMIC_MOVE)
+                // TODO: Not sure if this is the best way to trigger a refresh?
+                ApplicationManager.getApplication().invokeLater {
+                    VfsUtil.markDirtyAndRefresh(false, true, true, modcachePath.toFile())
```

Comment:
This appears to do the trick buf the read/write/EDT threading model is very complex to understand. I'm performing operations that are only available outside the VFS abstractions (like atomic move of temp dirs) and so to avoid issues need to figure out how best to notify IntelliJ of these changes.